### ISSUE TITLE
failed to run multi-GPU training on AN4 dataset

### ISSUE THREAD

**Issue by xhzhao** [SeanNaren/deepspeech.pytorch#279]
i tried to run multi-GPU training on AN4 dataset, but the program is blocked at the first iteration.
my command line is:
`python -m multiproc train.py  --rnn-type gru --hidden-size 800 --hidden-layers 5  --train-manifest data/an4_train_manifest.csv --val-manifest data/an4_val_manifest.csv --epochs 100 --batch-size 32 --learning-anneal 1.01  --cuda
`

Am i missing something?

pytorch version: 0.3.1.post2
the output log is: 

> ['train.py', '--rnn-type', 'gru', '--hidden-size', '800', '--hidden-layers', '5', '--train-manifest', 'data/an4_train_manifest.csv', '--val-manifest', 'data/an4_val_manifest.csv', '--epochs', '100', '--batch-size', '32', '--learning-anneal', '1.01', '--cuda', '--world-size', '4', '--rank', '0', '--gpu-rank', '0']
['train.py', '--rnn-type', 'gru', '--hidden-size', '800', '--hidden-layers', '5', '--train-manifest', 'data/an4_train_manifest.csv', '--val-manifest', 'data/an4_val_manifest.csv', '--epochs', '100', '--batch-size', '32', '--learning-anneal', '1.01', '--cuda', '--world-size', '4', '--rank', '1', '--gpu-rank', '1']
['train.py', '--rnn-type', 'gru', '--hidden-size', '800', '--hidden-layers', '5', '--train-manifest', 'data/an4_train_manifest.csv', '--val-manifest', 'data/an4_val_manifest.csv', '--epochs', '100', '--batch-size', '32', '--learning-anneal', '1.01', '--cuda', '--world-size', '4', '--rank', '2', '--gpu-rank', '2']
['train.py', '--rnn-type', 'gru', '--hidden-size', '800', '--hidden-layers', '5', '--train-manifest', 'data/an4_train_manifest.csv', '--val-manifest', 'data/an4_val_manifest.csv', '--epochs', '100', '--batch-size', '32', '--learning-anneal', '1.01', '--cuda', '--world-size', '4', '--rank', '3', '--gpu-rank', '3']
Model Save directory already exists.
DDP start
DDP end
DistributedDataParallel(
  (module): DeepSpeech(
    (conv): Sequential(
      (0): Conv2d(1, 32, kernel_size=(41, 11), stride=(2, 2), padding=(0, 10))
      (1): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True)
      (2): Hardtanh(min_val=0, max_val=20, inplace)
      (3): Conv2d(32, 32, kernel_size=(21, 11), stride=(2, 1))
      (4): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True)
      (5): Hardtanh(min_val=0, max_val=20, inplace)
    )
    (rnns): Sequential(
      (0): BatchRNN(
        (rnn): GRU(672, 800, bias=False, bidirectional=True)
      )
      (1): BatchRNN(
        (batch_norm): SequenceWise (
        BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
        (rnn): GRU(800, 800, bias=False, bidirectional=True)
      )
      (2): BatchRNN(
        (batch_norm): SequenceWise (
        BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
        (rnn): GRU(800, 800, bias=False, bidirectional=True)
      )
      (3): BatchRNN(
        (batch_norm): SequenceWise (
        BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
        (rnn): GRU(800, 800, bias=False, bidirectional=True)
      )
      (4): BatchRNN(
        (batch_norm): SequenceWise (
        BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
        (rnn): GRU(800, 800, bias=False, bidirectional=True)
      )
    )
    (fc): Sequential(
      (0): SequenceWise (
      Sequential(
        (0): BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True)
        (1): Linear(in_features=800, out_features=29, bias=False)
      ))
    )
    (inference_softmax): InferenceBatchSoftmax(
    )
  )
)
Number of parameters: 38067968
epoch start
iter start

my debug code:
![Uploading image.png…]()

**Comment by Wilbur529** [comment]
I couldn't see you debug code .. but i think we may get similar problem.
I tried to run the multi-gpus training these days, and also be failed. Could you show me more error information, and we could discuss it together ^_^

**Comment by SeanNaren** [comment]
@ggdupont @xhzhao do we still see this issue with 0.4/latest master?

**Comment by yrad** [comment]
I'm experiencing a similar issue using the multiproc module in the latest master.

Run in a container built from the dockerfile. Used the `--shm-size 8G` option to counter an _Unexpected bus error encountered in worker. This might be caused by insufficient shared memory_ error. Docker version 18.06.0-ce, build 0ffa825, nvidia-docker2

On a local machine, with default README params, training on a single GPU works fine, but multiproc stalls. Same log as @ggdupont. GPU util flips to %100 but memory and power usage is near idle. Rerunning with `--num-workers 0` made no difference, so I don't think it's related to the shm error.

benchmark.py fails with a _TypeError: initializer for ctype 'struct THIntTensor *' must be a pointer to same type, not cdata 'struct THFloatTensor *'_ but I don't think this is related.

**Comment by zinzinhust96** [comment]
@SeanNaren I still have this issue when running on a single GPU. The program is blocked in the first iteration when the parameter --cuda is passed in. It worked flawlessly without the option --cuda
